platforms/documentation/docs/src/docs/userguide/reference/best-practices/best_practices_general.adoc

```diff
@@ -142,6 +142,229 @@ include::sample[dir="snippets/bestPractices/usePluginsBlock-do/groovy",files="bu
+They can be difficult to decipher and may depend on configuration details that are hard to understand completely without running the build.
+
+Avoid build logic that assumes ordering between different projects, included builds, or applied scripts; the actual order is not guaranteed and is extremely vulnerable to minor structural changes (such as adding a new project or renaming an include).
+
+==== Plugin Developers
+
+Users should be able to apply your plugin in either order and have it behave correctly:
+
+====
+[.multi-language-sample]
+=====
+[source, kotlin]
+----
+plugins {
+  id("my-plugin")
+  id("plugin-i-depend-on")
+}
+----
+=====
+[.multi-language-sample]
+=====
+[source, groovy]
+----
+plugins {
+  id("my-plugin")
+  id("plugin-i-depend-on")
+}
+----
+=====
+====
+
+or
+
+====
+[.multi-language-sample]
+=====
+[source, kotlin]
+----
+plugins {
+  id("plugin-i-depend-on")
+  id("my-plugin")
+}
+----
+=====
+[.multi-language-sample]
+=====
+[source, groovy]
+----
+plugins {
+  id("plugin-i-depend-on")
+  id("my-plugin")
+}
+----
+=====
+====
+
+If your plugin only works in one of these cases, it’s relying on plugin order and will be fragile in real builds.
+
+If your plugin cannot function without another plugin, apply it explicitly at the start of `Plugin.apply`:
+
+====
+[.multi-language-sample]
+=====
+[source, kotlin]
+----
+// Ensure required plugin is applied
+project.pluginManager.apply("com.example.required-plugin")
+----
+=====
+[.multi-language-sample]
+=====
+[source, groovy]
+----
+// Ensure required plugin is applied
+project.pluginManager.apply('com.example.required-plugin')
+----
+=====
+====
+
+If your plugin only needs to integrate with another plugin when it’s present, react to its application using link:{javadocPath}/org/gradle/api/plugins/PluginManager.html#withPlugin(java.lang.String,org.gradle.api.Action)[`pluginManager.withPlugin()`] or link:{javadocPath}/org/gradle/api/DomainObjectCollection.html#configureEach(org.gradle.api.Action)[`plugins.configureEach {}`]:
+
+====
+[.multi-language-sample]
+=====
+[source, kotlin]
+----
+// Configure behavior that depends on required-plugin
+project.pluginManager.withPlugin("com.example.required-plugin") {  }
+
+project.plugins.configureEach { plugin ->
+    when (plugin) {
+        // Configure behavior specific to RequiredPlugin
+        is com.example.RequiredPlugin -> {  }
+        // Configure behavior specific to OtherPlugin
+        is com.example.OtherPlugin -> {  }
+    }
+}
+----
+=====
+[.multi-language-sample]
+=====
+[source, groovy]
+----
+// Configure behavior that depends on required-plugin
+project.pluginManager.withPlugin('com.example.required-plugin') {  }
+
+project.plugins.configureEach { plugin ->
+    // Configure behavior specific to RequiredPlugin
+    if (plugin instanceof com.example.RequiredPlugin) {  }
+    // Configure behavior specific to OtherPlugin
+    else if (plugin instanceof com.example.OtherPlugin) {  }
+}
```

Comment:
❗ I don't think we should include `configureEach`.  `withPlugin` is the preferred method in almost every case.  If we want to help people choose between alternatives we should firmly recommend that.

Reply:
I forgot about that case.  I think we should make it clear that if an id is present, you should use `withPlugin`.  This provides clear guidance of which alternative to choose most of the time.